충남대 BE_김해경 2주차 (2단계)

HaegyeongKim01 opened the pull request:
아직 SpringSecurirty를 꼭 구현할 필요는 없다고 하여 로그인 인증 수행 시 JWTTUtils에서 Jwt 토큰 생성 후 반환하는 방식으로 구현했는데 아직 다듬을 것이 많은 코드라고 생각합니다! 비밀번호 암호화 부분도 추후에 적용 해보려 합니다😃
- ViewController를 MemberController와 ProductController에서 따로 분리 하였는데 이렇게 사용해도 될까요??

HaegyeongKim01 committed:
feat: �¬�愿¢由 肄� 異�

HaegyeongKim01 committed:
feat: Service �대�� 異�

HaegyeongKim01 committed:
refactor: produc챙ct�� ProductDto濡 蹂¢寃

HaegyeongKim01 committed:
bug: service 二쇱� �ㅻ� �닿껐

HaegyeongKim01 committed:
feat: �¬�紐 寃¢利

HaegyeongKim01 committed:
validation �〈� 異�

HaegyeongKim01 committed:
[Step1] 湲곕 �뎄�ы 챙��¦

HaegyeongKim01 committed:
feat: 媛¢寃 �¦� �퓩≧� 寃¢�

HaegyeongKim01 committed:
Merge branch 'haegyeongkim01' into step1

HaegyeongKim01 committed:
style: BadRequestHandler 줄바꿈 추가

HaegyeongKim01 committed:
refactor: �¬�紐 �퓩≧� 寃¢� service 濡黝��¦ 泥━

HaegyeongKim01 committed:
[Step2] docs: 湲곕 �뎄�ы �¦

HaegyeongKim01 committed:
feat: �蕤��¦ 諛�� 媛庖껜 error 泥━

HaegyeongKim01 committed:
rename: index.html to admin.html

HaegyeongKim01 committed:
design: admin page

HaegyeongKim01 committed:
refactor: view 遺¦由

HaegyeongKim01 committed:
refactor: view 遺¦由

HaegyeongKim01 committed:
Merge branch 'step1' of https://github.com/HaegyeongKim01/spring-gift-wishlist into step2

HaegyeongKim01 committed:
feat: ���� �竈吏¢

HaegyeongKim01 committed:
feat: 濡帷렇� �竈吏¢

HaegyeongKim01 committed:
vifeat: ����, 濡帷렇� view 異�

HaegyeongKim01 committed:
feat: �� 媛庖껜, �ㅽ\ㅻ� 異�

HaegyeongKim01 committed:
feat: 濡帷렇� �몄�, �퓩¬ 諛�

HaegyeongKim01 committed:
feat: ���� 湲곕 援ы�

HaegyeongKim01 committed:
refactor: 濡帷렇� HTTP Status UNAUTHORIZED 蹂¢寃

HaegyeongKim01 committed:
Merge branch 'haegyeongkim01' into step2

Hyeon9mak commented on README.md:
````diff
@@ -1,6 +1,51 @@
+
+```
+
+### 구현 List
+
+- [ ]  User Schema.sql 작성
+- [ ]  UserDto 작성
+- [ ]  User 객체 클래스 작성
+- [ ]  인증하는 부분 작성
````
👀

Hyeon9mak commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@Controller
```
`@RestController` 를 사용해도 좋을 시점이네요!

Hyeon9mak commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,62 @@
+     * @param model Model
+     * @return JSON { "token" : ""}
+     */
+    @PostMapping("/login")
+    public ResponseEntity<Map<String, String>> login(LoginDto loginDto, Model model) {
+        Member member = loginDto.toUser();
+        try {
+            String token = service.login(member);
+
+            Map<String, String> response = new HashMap<>();
+            response.put("token", token);
+
+            return ResponseEntity.ok(response);
+        } catch (RuntimeException e) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+        }
+    }
```
controller advice 를 활용해 예외가 알아서 전달되게끔 하는 방법은 어떨까요?

HaegyeongKim01 replied:
아하~ 넵!

Hyeon9mak commented on src/main/java/gift/controller/ViewController.java:
```diff
@@ -26,4 +26,14 @@ public String index(Model model) {
         model.addAttribute("products", productList);
         return "admin";
     }
+
+    @GetMapping("/join")
+    public String join(Model model) {
+        return "join";
+    }
+
+    @GetMapping("/login")
+    public String login(Model model) {
+        return "login";
+    }
```
> ViewController를 MemberController와 ProductController에서 따로 분리 하였는데 이렇게 사용해도 될까요??

넵 물론이죠!
나누는 기준은 무엇이 될까요? 당연하게도 "이렇게 나누는게 나중에 유지보수에 유리할까? 편할까?" 입니다.
view page 만 자주 변경되는 상황이라면 지금처럼 분리해주는 것이 💯 
만약 menu api 가 바뀔 때 menu view page 도 함께 자주 바뀌는 상황이라면 합쳐두는게 유리하겠네요~

HaegyeongKim01 replied:
아하~~~ 넵 감사합니다~~

Hyeon9mak commented on src/main/java/gift/service/JwtUtil.java:
```diff
@@ -0,0 +1,58 @@
+@Component
+public class JwtUtil {
+
+    // Token 만료 시간
+    private final long TOKEN_TIME = 60 * 60 *1000L; //60분
+
```
주석을 통한 시간 단위 설명! 👍 
이런 사소한 배려가 동료들의 생산성에 큰 영향을 미치죠~

시간값도 static 하게 다룰 수 있겠네요.

HaegyeongKim01 replied:
넵 갑사합니당!

Hyeon9mak commented on src/main/java/gift/service/JwtUtil.java:
```diff
@@ -0,0 +1,58 @@
+    // Token 만료 시간
+    private final long TOKEN_TIME = 60 * 60 *1000L; //60분
+
+    private final static String SECRET_KEY = "mysecretmysecretmysecretmysecretmysecretmysecret";
+    private final static SecretKey key = Keys.hmacShaKeyFor(SECRET_KEY.getBytes());
```
상수는 대문자가 컨벤션 같아요 😄

HaegyeongKim01 replied:
아~ 넵 수정했습니다.

Hyeon9mak commented on src/main/java/gift/service/JwtUtil.java:
```diff
@@ -0,0 +1,58 @@
+     * @param role: 권한
+     * @return JWT 토큰 생성 후 반환
+     */
+    public String generateToken(String username, MemberRole role) {
+        return Jwts.builder()
+                .setSubject(username)
+                .claim("role", role)
+                .setExpiration(new Date(new Date().getTime() + TOKEN_TIME))
+                .signWith(key)
+                .compact();
+    }
```
expiration 도 있으니, date (발급일시)도 넣어주는건 어떨까요?

HaegyeongKim01 replied:
넵!

Hyeon9mak commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,41 @@
+        String email = member.getEmail();
+        Member foundMember = memberDao.findMemberByEmailAndPassword(email, member.getPassword());
+
+        if (foundMember == null || !foundMember.getEmail().equals(member.getEmail())) { // 검증 코드
+            throw new RuntimeException("이메일 혹은 비밀번호가 일치하지 않습니다.");
+        }
+        return createJwtToken(email, member.getRole());
```
1. rowMapper 로 생성절차를 거치기 때문에 foundMember 가 null 일 가능성은 없을거 같아요.
2. member 객체에게 이메일이 같은지, 다르다면 예외를 터치도록 질의하는건 어떨까요? 이러면 어떤 장점이 생길까요?

HaegyeongKim01 replied:
아! 그러네요 Member객체에서 이메일을 인증한다면 Member객체와 관련된 메소드 응집도가 향상되고 나중에 다른 곳에서 사용될 수 있도록 재사용성도 증가하는 장점이 존재할 것 같아요

Hyeon9mak commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,41 @@
+        return createJwtToken(email, member.getRole());
+    }
+
+    public String join(Member member) {
+        Boolean isSuccess = memberDao.addMember(member);
+        if (!isSuccess) {
+            throw new RuntimeException("회원가입에 실패했습니다.");
+        }
+        return login(member);
+    }
```
어떤 이유로 회원가입에 실패했을까요? 사용자가 답답함을 느낄 수도 있을거 같아요.

HaegyeongKim01 replied:
아! 그럼 구현은 Dao 클래스에서 DB에 쿼리 날릴 때 예외가 발생하면 try catch로 처리하는게 좋을까요? 조언 부탁드립니다!

Hyeon9mak replied:
아, 코멘트의 의도는 더욱 상세한 예외 메세지를 남겨주는 것이 어떠한가! 였어요
에러 메세지는 어떤 의미를 담는 것이 좋을까요?
https://hyeon9mak.github.io/writing-helpful-error-messages/

Hyeon9mak commented on src/main/java/gift/vo/Member.java:
```diff
@@ -0,0 +1,29 @@
+    private String password;
+    private MemberRole role;
+
+    public Member(String email, String password) {
+        this(email, password, null);
+    }
+
+    public Member(String email, String password, MemberRole role) {
+        this.email = email;
+        this.password = password;
+        this.role = role;
+    }
```
null 한 role 대신, 기본 role 을 지정해주는 건 어떨까요? 😉

HaegyeongKim01 replied:
아하! 넵 좋은 것 같아요

Hyeon9mak reviewed (changes requested):
안녕하세요 해경님! 2단계도 잘 진행해주셨습니다 👍 
이정도 규모를 구현하는데 꼭 spring security 를 사용할 필요는 없죠.
배보다 배꼽이 더 큰 프로젝트가 될 수 있으니까요~
간단한 코멘트 남겼으니 확인해주세요!

HaegyeongKim01 committed:
docs: 援ы� �¦猷� 湲곕 list ��

HaegyeongKim01 committed:
refactor: 蹂¢寃�Controller to @RestController

HaegyeongKim01 committed:
refactor: Exception handler ��

HaegyeongKim01 committed:
refactor: Member default 沅� 異�

HaegyeongKim01 committed:
refactor: �퓩¬곗� 諛帷��쇱 異�

HaegyeongKim01 committed:
refactor: 蹂¢� static 蹂¢寃

HaegyeongKim01 committed:
refactor: 以¦諛� �帷굅

HaegyeongKim01 committed:
refactor: Member媛庖껜�¦ �대�� �몄�

�대�� 寃¢利 濡黝�� Member �대�� �대�濡 �대�� 愿¢� 濡黝�� � 怨녹�
紐⑥鯛�.
�ъ\ъK� 利�
媛¢�� 媛黝¦
�¦� �μ�� �린�¦濡 肄� 蹂¢寃

Hyeon9mak commented on src/main/java/gift/vo/Member.java:
```diff
@@ -0,0 +1,39 @@
+        return role;
+    }
+
+    /**
+     * 객체와 매개변수로 받은 email이 일치하는지 검증하는 메소드
+     * @param email 검증할 email
+     */
+    public void validateEmail(String email) {
+        if (!this.email.equals(email)) {
+            throw new RuntimeException("이메일");
+        }
+    }
```
역시나 상세한 메세지를 담아주는게 좋겠네요~
관련 테스트 코드도 작성되면 좋겟어요!

Hyeon9mak commented on src/main/java/gift/vo/Member.java:
```diff
@@ -0,0 +1,39 @@
+    private String password;
+    private MemberRole role;
+
+    public Member(String email, String password) {
+        this(email, password, String.valueOf(MemberRole.USER));
+    }
```
😉 👍

Hyeon9mak commented on src/main/resources/db/h2/schema.sql:
```diff
@@ -1,10 +1,18 @@
+);
+
+-- -- Create the USER table
+CREATE TABLE MEMBER (
+                        email VARCHAR(100) PRIMARY KEY,
+                        password VARCHAR(100) NOT NULL,
+                        role VARCHAR(20)
```
이제 role 또한 NOT NULL 하게 관리할 수 있겠네요~

HaegyeongKim01 replied:
아~ 그러네요 필드 수정 후 데이터 스키마도 확인하는 습관을 가지는 것이 좋을 것 같다고 느꼈습니다~!

Hyeon9mak reviewed (approved):
안녕하세요 해경님! 피드백 빠르게 잘 반영해주셨어요 👍 
다음 단계 진행에 참고하실만한 코멘트 남겼습니다.
확인 후 다음 단계 진행해주세요!
